### ISSUE TITLE
Running `docker build` from a script

### ISSUE THREAD

**Issue by remram44** [docker/docker#7727]
Hi,

I'm building a Docker container from a script. However:
* `docker build` has no option that will give you the image id in a readable format; best you can do is parse the whole output on stdout with a regex.
* `docker build`'s output really should be on stderr
* I was expecting `--quiet` to suppress that output, but it only removes the output from commands run in containers, not the rest of the client's output.

What I was expecting is all the current output to happen on stderr, unless an option silences it (`--quiet` is already taken, but that's really what it should do), and the new image id to be printed on stdout for script consumption (`my_shiny_image=$(docker build . 2>/dev/stderr)`).

The best I can do right now is redirect stderr and stdout to /dev/null to suppress them, and generate a unique tag name myself to pass in with `-t` (parsing stdout just doesn't make sense).

(version: 1.0.0, 63fe64c)

**Comment by ripcurld00d** [comment]
#dibs

**Comment by boxofrox** [comment]
Just an FYI.  @creack commented on stackoverflow [1] that the build log should go out on stderr and the image ID on stdout.  I intended to file a bug as suggested, and found this issue coincided.

While I'm happy with a command line flag that dumps the ID, I'm curious why the build log didn't remain on stderr.

[1] http://stackoverflow.com/questions/22080706/how-to-create-named-and-latest-tag-in-docker/22081272#comment38742167_22081272

**Comment by ripcurld00d** [comment]
@boxofrox thanks for notifying me about this. I didn't know that.

Looking at the current [api/client/build.go](https://github.com/docker/docker/blob/master/api/client/build.go), it looks like the daemon is using `cli.out` to stream its output. However, it has been like that from [the beginning](https://github.com/docker/docker/blob/58690c9cca5995035b721ed6f1337b0ddb932d7a/api/client/build.go). So, I'm probably searching at the wrong place. Maybe someone else can answer this question?

**Comment by boxofrox** [comment]
@ripcurld00d found it.  CLI commands used to be in one file `api/client/commands.go` before they were split in Mar 2015 (from the beginning as you said ;) ).  commit 5c91bb9.

Seems to me like the fix at the time should have been to disable output on `cli.err` when the `--quiet` flag was given instead of switching build output to `cli.out`.

**Comment by boxofrox** [comment]
More FYI.  As a linux user, I use scripts to automate/customize tedious tasks with command line tools.  I've found [Hints for writing Unix tools](http://monkey.org/~marius/unix-tools-hints.html) inspirational when writing my own CLI programs, particularly the bits about stdout being succint and effectively an API, while stderr is for errors, debugging, and misc info like the build log in this issue.

**Comment by ripcurld00d** [comment]
@boxofrox :+1: Good job! Thanks.

**Comment by duglin** [comment]
@ripcurld00d @boxofrox I'd be curious to know what you thought about these semantics:
1 - no `-s` -> keep today's output, most everything goes to stdout, final error msg, if any, goes to stderr
2 - `-s` + build error -> stdout is empty, stderr is full build output
3 - `-s` + successful build -> stdout has image ID, stderr is empty

My reasoning for having the stderr be empty in the third case is because it feels a bit odd to have a "silent" flag that still allows for output to appear, even if its on stderr, in the successful case.  However, if there is an error then I think showing it (on stderr) makes some sense.  However, I do tend to agree with the comment above that it would be nice if even w/o -s the build output went to stderr and just the final ID appeared on stdout, but I think that's a bigger/breaking change that might be harder to get thru.

**Comment by ripcurld00d** [comment]
@duglin @boxofrox :+1: (fine with me).

**Comment by boxofrox** [comment]
@duglin 

(2) doesn't seem like anything is silenced.  If I were to write my script relying on (3), I'd always be redirecting my stderr to `/dev/null` if I wanted to silence `docker build`'s "misc" output.  Other than that quibble, I can work with that. 

@duglin, @ripcurld00d 

I understand that docker's CLI may run into difficult technical problems, especially supporting an HTTP interface, and won't be an "ideal" unix command line tool.  And breaking existing behavior is definitely a valid concern.

Ideally, IMHO, the silent flag `-s` would be the equivalent of redirecting stderr to null `2>/dev/null`.  Presence of errors is indicated primarily by a non-zero return value, and the image ID is the only output on stdout.

It looks like there are three types of output here: the image ID, the build log, and the final error message.  And we only have two channels for output: stdout, stderr.  Image ID belongs only on stdout, final error message belongs only on stderr, but where to put the build log?  Some want it on stdout, others and myself prefer it on stderr.

Perhaps it makes sense to have an output flag.
```sh
docker build --output=unix-cli ...
docker build --output=logger ...
```
- `--output=unix-cli` would put the build log and final error message on stderr, and `-s` just silences stderr.  only image ID or nothing appears on stdout.
- `--output=logger` would put the build log on stdout, with final error message on stderr.  I dunno what `-s` would do in this case (maybe silence stderr?).
- if `--output` is omitted, it defaults to `logger`.

I'm just spitballing here, more so to just consider other possibilities that makes everyone happy, so I'm not adamant about this proposal.  There are obvious downsides:

1. adds another cli flag.
2. other cli flags `-s` could behave differently based on `--output`'s value unless restrained by policy.
3. the `build` code may have a severe split personality disorder.

The one pro I can think of is that if I want the build log + final error on stderr, it's available to me.

Actually, this proposal sounds like @ripcurld00d's original intent with `--silent` switching to script mode.  Maybe he just had the wrong flag name :stuck_out_tongue_winking_eye:

Sorry for the really long post.

**Comment by duglin** [comment]
@boxofrox  said:
- - - - - - -
(2) doesn't seem like anything is silenced. If I were to write my script relying on (3), I'd always be redirecting my stderr to /dev/null if I wanted to silence docker build's "misc" output. Other than that quibble, I can work with that. 
- - - - - - - -

In the successful build case the redirect wouldn't be needed since nothing would be sent to stderr, so things are silenced but just when the build worked.

Let me ask you this.... in the case where the build fails would you be more annoyed with having to rerun the build again just to see the error message, or would you be more annoyed with the error messages being there at all? I'm not totally thrilled with `-s` not actually being silent in (2) but the more I thought about it the more it seemed to make sense from a UX perspective.  If something goes wrong then we're asking people to rerun the command all over again w/o the -s just to see the error message.  Seems kind of annoying.  However, I do agree that if "-s" means "silent" then (like curl) perhaps it really should be.  Of course, even then we're not totally silent because of the ID.  :-)   I did struggle with this as I was writing up the previous comment and in the end I landed on "-s" lying a little about how silent it is.  But I can be convinced to make it totally silent (modulo the ID).  We're in kind of a bad spot due to the way it was originally designed.

Not too excited about adding an "output" flag - seems like it might be overkill.

**Comment by boxofrox** [comment]
> In the successful build case the redirect wouldn't be needed since nothing would be sent to stderr, so things are silenced but just when the build worked.

There's no way for my bash script to know in advance whether a build will fail so it could redirect to `/dev/null`.  Hence, it's always there if I'm certain I don't want to see output from stderr.

> Let me ask you this.... in the case where the build fails would you be more annoyed with having to rerun the build again just to see the error message, or would you be more annoyed with the error messages being there at all?

I definitely don't want to run the build again.  That seems like a harsh precedent to impose on fellow scripters, so I won't advocate it.  I'm honestly more concerned with the build log being on stderr than a `--silent` flag.  Once there, I can decide with redirection how silent I want to be, and I'd always get an image ID on stdout I can stash in a variable without resorting to fragile `grep`/`awk`/`sed` regexes.

> Of course, even then we're not totally silent because of the ID.

This did cross my mind.  I think it varies from tool to tool.  When you only care about the return value, total silence is preferred.  Since `docker build` can take quite some time to complete, I'd prefer to always have an image ID come back, otherwise we might need separate silent flags, or silent modes.  It's definitely a tradeoff.

So, in conclusion, I can work with your 3 point proposal.  It gets me (and @remram44 I believe) where we want to be: only image ID on stdout.

**Comment by virtuald** [comment]
`docker run` has a `--cidfile` argument. It's a bit strange that `docker build` doesn't have a similar thing, if it did then we wouldn't need to worry as much about how silent or not silent the output was.

**Comment by thaJeztah** [comment]
@ripcurld00d opened a PR to make `docker build -q` (quiet) only output the image-id on success; https://github.com/docker/docker/pull/17428, so this will be resolved in docker 1.10

@virtuald could be useful, but I think we should have a separate issue for that

**Comment by virtuald** [comment]
@thaJeztah it's related to this issue, as a `--cidfile` option is something that would really only be useful for scripting

**Comment by ripcurld00d** [comment]
I agree with @thaJeztah.
@virtuald - it is not really related to this issue. However, it is a nice-to-have feature.

I will write a proposal for it in the following days, though.

Thanks.

**Comment by ripcurld00d** [comment]
@thaJeztah -- #17428 closes this too. Thanks :)

**Comment by thaJeztah** [comment]
@ripcurld00d oh, yes!
